src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1252,5 +1245,236 @@ internal bool CanAddParsedValue(HttpHeaderParser parser)
+
+                int count = _count;
+                _count++;
+                if ((uint)count < (uint)entries.Length)
+                {
+                    entries[count].Key = key;
+                    return ref entries[count].Value!;
+                }
+
+                return ref GrowEntriesAndAddDefault(key);
+            }
+            else if (store is null)
+            {
+                _count++;
+                entries = new HeaderEntry[InitialCapacity];
+                _headerStore = entries;
+                ref HeaderEntry firstEntry = ref MemoryMarshal.GetArrayDataReference(entries);
+                firstEntry.Key = key;
+                return ref firstEntry.Value!;
+            }
+            else
+            {
+                return ref DictionaryGetValueRefOrAddDefault(key);
+            }
+
+            ref object? GrowEntriesAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                if (entries.Length == ArrayThreshold)
+                {
+                    return ref ConvertToDictionaryAndAddDefault(key);
+                }
+                else
+                {
+                    Array.Resize(ref entries, entries.Length << 1);
+                    _headerStore = entries;
+                    ref HeaderEntry firstNewEntry = ref entries[entries.Length >> 1];
+                    firstNewEntry.Key = key;
+                    return ref firstNewEntry.Value!;
+                }
+            }
+
+            ref object? ConvertToDictionaryAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                var dictionary = new Dictionary<HeaderDescriptor, object>(ArrayThreshold);
+                _headerStore = dictionary;
+                foreach (HeaderEntry entry in entries)
+                {
+                    dictionary.Add(entry.Key, entry.Value);
+                }
+                Debug.Assert(dictionary.Count == _count - 1);
+                return ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
+            }
+
+            ref object? DictionaryGetValueRefOrAddDefault(HeaderDescriptor key)
+            {
+                var dictionary = (Dictionary<HeaderDescriptor, object>)_headerStore!;
+                ref object? value = ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
+                if (value is null)
+                {
+                    _count++;
+                }
+                return ref value;
+            }
+        }
+
+        private void AddEntryToStore(HeaderEntry entry)
+        {
+            Debug.Assert(!ContainsKey(entry.Key));
+
+            if (_headerStore is HeaderEntry[] entries)
+            {
+                int count = _count;
+                if ((uint)count < (uint)entries.Length)
```

Comment:
Why the casts here? Similarly elsewhere.

Reply:
It allows the JIT to avoid inserting a length check on the `entries[count]` access.

Reply:
Is this worthwhile to do?

If it is, let's add a comment. Otherwise it will likely get deleted in a future PR.

Reply:
I don't have numbers on exactly how much it contributes here, but together these reduce the amount of generated code by quite a bit. I'm sure there are lots of places throughout runtime we use this trick that are less perf sensitive than this though.

Reply:
It is very common across the codebase. Especially in corelib.